### PR TITLE
surface: redesign state

### PR DESCRIPTION
**Pull request by emersion**
- [x] Rename `invalid` to `committed` - better ideas welcome!
- [x] Move `subsurface_position` in `wlr_subsurface`
- [x] Fix `sx`/`sy` handling
- [x] Don't monkey-patch `surface_damage` and `buffer_damage` - add a new buffer damage region instead
- [x] Make `current` and `pending` embedded structs
- [x] Add a `previous` state
- [x] Update cursor hotspot with buffer position

Test plan:
- Make sure damage tracking works on surfaces and subsurfaces (try gnome-calculator)
- Try with scaled and transformed outputs
- Try [wleird](https://github.com/emersion/wleird)'s cursor example

See #1065

### CODE REVIEW

**Comment by ammen99** [comment]
> Remove subsurface_position

Is there another way to get the subsurface position relative to its parent?

**Comment by emersion** [comment]
>Is there another way to get the subsurface position relative to its parent?

Yes, this will be moved in `wlr_subsurface`.

File: include/wlr/types/wlr_surface.h [hunk 1]
```diff
@@ -8,57 +8,32 @@
...
-#define WLR_SURFACE_INVALID_OPAQUE_REGION 8
-#define WLR_SURFACE_INVALID_INPUT_REGION 16
-#define WLR_SURFACE_INVALID_TRANSFORM 32
-#define WLR_SURFACE_INVALID_SCALE 64
-#define WLR_SURFACE_INVALID_SUBSURFACE_POSITION 128
-#define WLR_SURFACE_INVALID_FRAME_CALLBACK_LIST 256
+enum wlr_surface_state_field {
+	WLR_SURFACE_STATE_BUFFER = 1,
+	WLR_SURFACE_STATE_SURFACE_DAMAGE = 2,
+	WLR_SURFACE_STATE_BUFFER_DAMAGE = 2,
```
Comment (synaptiko) [inline include/wlr/types/wlr_surface.h]: Should it be `4` here?

Comment (emersion) [reply include/wlr/types/wlr_surface.h]: Right, thanks!

**Comment by emersion** [comment]
I'm not really happy with the current `wlr_surface_state` struct. It contains both things from the client (fields from `buffer` to `frame_callback_list`) accumulated before commit and things computed on commit (`width`, `height`, `s{x,y}`, and I wanted to add the `wlr_buffer`?).

**Comment by emersion** [comment]
Still not sure how damage should be defined.
* What should be the (0, 0) point in the damage region? The top-left corner of the buffer or the top-left corner of the surface?
* Should it be in buffer-local coords or in surface-local coords? We loose precision if we do surface-local coords, and it's doesn't add a _lot_ of complexity in the damage code - just one `/ surface->current.scale` somewhere, and one `wlr_transform_compose(…, surface->current.transform)` somewhere else.

**Comment by ammen99** [comment]
@emersion, as we're doing the `damage` for just convenience, I think it should be in surface local coordinates, because that's 
what a compositor typically uses. In this case it'll also make sense to make (0,0) the top left corner of the surface. I'm however not sure how to do precision issues, maybe `wlr_region_expand`?

**Comment by SirCmpwn** [comment]
It should be buffer local.

**Comment by ammen99** [comment]
> It should be buffer local.

I'm not trying to argue, just asking out of interest: what makes you think this way?

**Comment by SirCmpwn** [comment]
Can we just go with "it's been a long day and explaining the full answer is beyond the limits of my patience right now"?

**Comment by emersion** [comment]
Moved bits that are computed as a delta of two states (the enhanced damage, the surface position, the buffer itself) outside of `wlr_surface_state`.

Decided to use damage in buffer-local coordinates, so that you can apply scale and transform to this region in the exact same way you render the associated texture.

**Comment by emersion** [comment]
This PR is ready for review.

**Comment by ammen99** [comment]
So, after this, what is the intended solution for #1065?

**Comment by emersion** [comment]
>So, after this, what is the intended solution for #1065?

Let's continue to discuss this in the other issue.

**Comment by SirCmpwn** [comment]
https://github.com/swaywm/sway/pull/2185

Code looks good, running this as my daily driver for a while.

**Comment by atomnuker** [comment]
Pointer position is still off after applying this and the sway update PR. Do you compensate for scaling?

**Comment by emersion** [comment]
>Pointer position is still off after applying this and the sway update PR. Do you compensate for scaling?

Input doesn't deal with scaling and transform. Can you describe how to reproduce? Are you sure you're running with the latest version of this branch? (A previous commit fucked up the input region)

**Comment by atomnuker** [comment]
Yep, this is with the latest version of this PR. There's no transform on anything, just scaling set in sway.
You've reproduced this before and I have mentioned this before too. Hotspot is fine, the image matches what's on the screen and sent as input, its just that if you open a right click menu its offset compared to the pointer location.

**Comment by emersion** [comment]
FYI I have input issues with some floating windows on Sway master (without this patch applied): https://github.com/swaywm/sway/issues/2195

**Comment by atomnuker** [comment]
In GTK - a few pixels (3 or 4) upwards and a pixel or two to the left of where the menu is created.
In xwayland (such as firefox), 6 or 7 pixels upwards (or downwards, if you try to create the menu too far down the screen) and about 1 or two pixels to the left.
Yes, this isn't a regression, its always been present, but since this commit touched some position code I thought it might also fix this issue.

**Comment by emersion** [comment]
>Yes, this isn't a regression, its always been present, but since this commit touched some position code I thought it might also fix this issue.

I think this is more of a xdg-shell window geometry issue (this PR only updates the wl_surface implementation).

**Comment by emersion** [comment]
Bump: I've been using this for a few days.

**Comment by SirCmpwn** [comment]
aight

**Pull request merged**